File: 02/solution.rb
test = [
  "forward 5",
  "down 5",
  "forward 8",
  "up 3",
  "down 8",
  "forward 2"
]

input_file = File.open('./input.txt', 'r')
input = input_file.readlines()
input_file.close

class Submarine
  attr_accessor :depth
  attr_accessor :position

  def initialize
    @depth = 0
    @position = 0
  end

  def forward(x)
    @position+=x
  end

  def up(x)
    @depth-=x
  end

  def down(x)
    @depth+=x
  end

  def move(instruction)
    parts = instruction.split(' ')
    message = parts[0]
    amount = parts[1].to_i
    send(message, amount)
  end
end

def part_one(input)
  submarine = Submarine.new
  input.each{ |i| submarine.move(i) }

  submarine.depth * submarine.position
end

puts part_one(input)

class SecondSubmarine
  attr_accessor :depth
  attr_accessor :position
  attr_accessor :aim

  def initialize
    @depth = 0
    @position = 0
    @aim = 0
  end

  def forward(x)
    @position+=x
    @depth+=@aim*x
  end

  def up(x)
    @aim-=x
  end

  def down(x)
    @aim+=x
  end

  def move(instruction)
    parts = instruction.split(' ')
    message = parts[0]
    amount = parts[1].to_i
    send(message, amount)
  end
end

def part_two(input)
  submarine = SecondSubmarine.new
  input.each{ |i| submarine.move(i) }

  submarine.depth * submarine.position
end

puts part_two(input)

File: 02/solution.exs
defmodule Input do
  def lines(filename) do
    {:ok, contents} = File.read(filename)
    contents
    |> String.split("\n")
  end
end

defmodule Submarine do
  def process([line | tail], state) when line == "", do: process(tail, state)

  def process(
    [line | rest],
    state
  ) do
    [instruction, x] = String.split(line) 
    process(rest, handle(state, instruction, String.to_integer(x)))
  end

  def process([], state), do: state

  def handle(state, "forward", x) do
    {_, new_state} = Map.get_and_update(state, :position, fn position ->
      {position, position+x}
    end)
    new_state
  end

  def handle(state, "up", x) do
    {_, new_state} = Map.get_and_update(state, :depth, fn depth ->
      {depth, depth-x}
    end)
    new_state
  end

    def handle(state, "down", x) do
    {_, new_state} = Map.get_and_update(state, :depth, fn depth ->
      {depth, depth+x}
    end)
    new_state
  end
end

defmodule Answer do
  def get(%{ depth: depth, position: position }), do: depth*position
end

Submarine.process(Input.lines("input.txt"), %{ depth: 0, position: 0 })
|> Answer.get
|> IO.puts

defmodule SecondSubmarine do
  def process([line | tail], state) when line == "", do: process(tail, state)

  def process(
    [line | rest],
    state
  ) do
    [instruction, x] = String.split(line) 
    process(rest, handle(state, instruction, String.to_integer(x)))
  end

  def process([], state), do: state

  def handle(state, "forward", x) do
    %{ aim: aim } = state
    state
    |> Map.update!(:position, &(&1+x))
    |> Map.update!(:depth, &(&1 + aim*x))
  end

  def handle(state, "up", x) do
    state
    |> Map.update!(:aim, &(&1 - x))
  end

  def handle(state, "down", x) do
    state
    |> Map.update!(:aim, &(&1 + x))
  end
end


SecondSubmarine.process(Input.lines("input.txt"), %{ depth: 0, position: 0, aim: 0 })
|> Answer.get
|> IO.puts

